spi_nand_flash/src/nand_diag_api.c

```diff
@@ -0,0 +1,158 @@
+    ESP_RETURN_ON_FALSE(temp_buf != NULL, ESP_ERR_NO_MEM, TAG, "nomem");
+
+    fill_buffer(PATTERN_SEED, pattern_buf, sector_size / sizeof(uint32_t));
+
+    int64_t read_time = 0;
+    int64_t write_time = 0;
+
+    for (int i = start_sec; i < (start_sec + sec_count); i++) {
+        int64_t start = esp_timer_get_time();
+        if (get_raw_tp) {
+            ESP_ERROR_CHECK(nand_wrap_prog(flash, i, pattern_buf));
+        } else {
+            ESP_ERROR_CHECK(spi_nand_flash_write_sector(flash, pattern_buf, i));
+        }
+        write_time += esp_timer_get_time() - start;
+
+        memset((void *)temp_buf, 0x00, sector_size);
+
+        start = esp_timer_get_time();
+        if (get_raw_tp) {
+            ESP_ERROR_CHECK(nand_wrap_read(flash, i, 0, sector_size, temp_buf));
+        } else {
+            ESP_ERROR_CHECK(spi_nand_flash_read_sector(flash, temp_buf, i));
+        }
+        read_time += esp_timer_get_time() - start;
+    }
+    free(pattern_buf);
+    free(temp_buf);
+
+    ESP_LOGI(TAG, "Wrote %" PRIu32 " bytes in %" PRId64 " us, avg %.2f kB/s", sector_size * sec_count, write_time, (float)sector_size * sec_count / write_time * 1000);
+    ESP_LOGI(TAG, "Read %" PRIu32 " bytes in %" PRId64 " us, avg %.2f kB/s\n", sector_size * sec_count, read_time, (float)sector_size * sec_count / read_time * 1000);
+    return ret;
+}
+
+static bool s_is_ecc_exceed_threshold(spi_nand_flash_device_t *handle)
+{
+    uint8_t min_bits_corrected = 0;
+    bool ret = false;
+    if (handle->chip.ecc_data.ecc_corrected_bits_status == STAT_ECC_1_TO_3_BITS_CORRECTED) {
+        min_bits_corrected = 1;
+    } else if (handle->chip.ecc_data.ecc_corrected_bits_status == STAT_ECC_4_TO_6_BITS_CORRECTED) {
+        min_bits_corrected = 4;
+    } else if (handle->chip.ecc_data.ecc_corrected_bits_status == STAT_ECC_7_8_BITS_CORRECTED) {
+        min_bits_corrected = 7;
+    }
+
+    if (min_bits_corrected >= handle->chip.ecc_data.ecc_data_refresh_threshold) {
+        ret = true;
+    }
+    return ret;
+}
+
+esp_err_t nand_get_ecc_stats(spi_nand_flash_device_t *flash)
+{
+    esp_err_t ret = ESP_OK;
+    uint32_t sector_size, block_size, num_blocks;
+    uint32_t ecc_err_total_count = 0;
+    uint32_t ecc_err_exceeding_threshold_count = 0;
+    uint32_t ecc_err_not_corrected_count = 0;
+
+    spi_nand_flash_get_sector_size(flash, &sector_size);
+    spi_nand_flash_get_block_size(flash, &block_size);
+    spi_nand_flash_get_block_num(flash, &num_blocks);
+    uint32_t pages_per_block = block_size / sector_size;
+    uint32_t num_pages = num_blocks * pages_per_block;
+
+    bool is_free = true;
+    for (uint32_t page = 0; page < num_pages; page++) {
+        ret = nand_wrap_is_free(flash, page, &is_free);
+        if (!is_free) {
+            ret = nand_wrap_get_ecc_status(flash, page);
+            if (ret) {
+                ESP_LOGE(TAG, "Failed to read ecc error for page=%" PRIu32 "", page);
+                return ESP_FAIL;
```

Comment:
Why don't we return 'ret' as given by nand_wrap_get_ecc_status()? Btw, using `if (ret) ` for esp_err_t looks confusing to me, as it passes for ret != ESP_OK but it's optically saying the other way round.

I just think that examples should be super-clear, sort of model application

Reply:
Updated. Thank you for raising this point.